llms/mlx_lm/utils.py

```diff
@@ -150,20 +189,27 @@ def generate(
         print("=" * 10)
         print("Prompt:", prompt)
 
-    prompt = mx.array(tokenizer.encode(prompt))
+    prompt_tokens = mx.array(tokenizer.encode(prompt))
+    repetition_context = []
+    repetition_context.extend(tokenizer.encode(prompt))
 
     tic = time.perf_counter()
     tokens = []
     skip = 0
     REPLACEMENT_CHAR = "\ufffd"
 
-    for (token, prob), n in zip(generate_step(prompt, model, temp), range(max_tokens)):
-        if token == tokenizer.eos_token_id:
+    for (token, prob), n in zip(
+        generate_step(
+            prompt_tokens, model, temp, repetition_penalty, repetition_context
+        ),
+        range(max_tokens),
+    ):
+        if token[0] == tokenizer.eos_token_id:
```

Comment:
What is the reason we need to change to use `token[0]` instead of `token`?

Reply:
In the `generate_step` function, I was squeezing the first dimension of the logits array, and I ended up getting a  array with topk tokens after `mx.argmax`, and `token==tokenizer.eos_token_id` would fail because `token` consisted of topk tokens. I thought `token[0]` would extract the actual value from the y tensor in `generate_step`.

Following up with your suggestion, I changed `logins.squeeze(1)` to `logins[ : , -1 , : ]` that allowed me to get the last tokens from the sequence dimension, solving this problem. Thanks for that :)